File: BackEnd/src/domain/usecases/weather/GetCurrentWeatherUseCase.ts
/**
 * @file GetCurrentWeatherUseCase.ts
 * @description Caso de uso para obtener el clima actual de una localización.
 * Delega en WeatherService la lógica de caché + API.
 * @module Weather
 * @layer Domain
 *
 * @implements {IGetCurrentWeatherUseCase}
 * @injectable
 * @dependencies WeatherService
 */

import { injectable, inject } from 'inversify';
import { TYPES } from '../../../core/types.js';
import type { IGetCurrentWeatherUseCase } from '../../interfaces/usecases/weather/IGetCurrentWeatherUseCase.js';
import type { WeatherQueryRequestDTO } from '../../dtos/weather/weather-query-request.dto.js';
import type { WeatherResponseDTO } from '../../dtos/weather/weather-response.dto.js';
import { WeatherService } from '../../../presentation/services/WeatherService.js';

/**
 * Obtiene el clima actual para las coordenadas dadas.
 *
 * @implements {IGetCurrentWeatherUseCase}
 * @injectable
 * @dependencies WeatherService
 */
@injectable()
export class GetCurrentWeatherUseCase implements IGetCurrentWeatherUseCase {
  constructor(
    @inject(TYPES.WeatherService) private readonly weatherService: WeatherService,
  ) {}

  /**
   * @param dto — Localización como string libre (`location`) o coordenadas (`lat`/`lon`).
   *              Si se proporciona `location`, tiene prioridad sobre lat/lon.
   * @returns WeatherResponseDTO con las condiciones actuales.
   * @throws {ExternalServiceException} Si la API no está disponible.
   */
  async execute(dto: WeatherQueryRequestDTO): Promise<WeatherResponseDTO> {
    if (dto.location?.trim()) {
      return this.weatherService.getWeatherForQuery(dto.location.trim());
    }
    return this.weatherService.getWeatherForLocation(dto.lat!, dto.lon!);
  }
}

File: FrontEnd/lib/domain/usecases/weather/get_current_weather_use_case.dart
/// @file get_current_weather_use_case.dart
/// @description Caso de uso para obtener el clima actual de una ubicación.
/// Delega en IWeatherRepository — sin lógica de negocio adicional más allá
/// de delegar la orquestación de caché y red al repositorio.
/// @module Weather
/// @layer Domain
library;

import '../../entities/weather_data.dart';
import '../../interfaces/usecases/weather/i_get_current_weather_use_case.dart';
import '../../repositories/i_weather_repository.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// GET CURRENT WEATHER USE CASE
// ═══════════════════════════════════════════════════════════════════════════════

/// Caso de uso que obtiene el clima actual para una ubicación dada.
///
/// Devuelve [WeatherData] con temperatura, condición, humedad y viento.
/// El caché y el modo mock son responsabilidad del repositorio.
///
/// [implements] IGetCurrentWeatherUseCase
/// [injectable] registrar en container.dart como lazySingleton.
/// [dependencies] IWeatherRepository.
class GetCurrentWeatherUseCase implements IGetCurrentWeatherUseCase {
  final IWeatherRepository _repository;

  const GetCurrentWeatherUseCase({required IWeatherRepository repository})
      : _repository = repository;

  /// Obtiene el clima actual para [location].
  ///
  /// [location] — nombre de ciudad o coordenadas aceptadas por WeatherAPI.com.
  /// [returns]  — [WeatherData] con los datos meteorológicos actuales.
  /// [throws]   AppError.network si no hay conexión y no hay caché válido.
  @override
  Future<WeatherData> execute(String location) {
    return _repository.getCurrentWeather(location);
  }
}
